community/bolt/src/main/java/org/neo4j/bolt/transport/TransportWriteThrottle.java

```diff
@@ -0,0 +1,106 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.transport;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandler;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.WriteBufferWaterMark;
+import io.netty.util.AttributeKey;
+
+import java.util.function.Supplier;
+
+/**
+ * Throttle that blocks write operations to the channel based on channel's isWritable
+ * property. Buffer sizes based on which the channel will change its isWritable property
+ * and whether to apply this throttle are configurable through GraphDatabaseSettings.
+ */
+public class TransportWriteThrottle implements TransportThrottle
+{
+    private static final AttributeKey<ThrottleLock> LOCK_KEY = AttributeKey.valueOf( "BOLT.WRITE_THROTTLE.LOCK" );
+    private final int lowWaterMark;
+    private final int highWaterMark;
+    private final Supplier<ThrottleLock> lockSupplier;
+    private final ChannelInboundHandler listener;
+
+    public TransportWriteThrottle( int lowWaterMark, int highWaterMark )
+    {
+        this( lowWaterMark, highWaterMark, () -> new DefaultThrottleLock() );
+    }
+
+    public TransportWriteThrottle( int lowWaterMark, int highWaterMark, Supplier<ThrottleLock> lockSupplier )
+    {
+        this.lowWaterMark = lowWaterMark;
+        this.highWaterMark = highWaterMark;
+        this.lockSupplier = lockSupplier;
+        this.listener = new ChannelStatusListener();
+    }
+
+    @Override
+    public void install( Channel channel )
+    {
+        ThrottleLock lock = lockSupplier.get();
+
+        channel.attr( LOCK_KEY ).set( lock );
+        channel.config().setWriteBufferWaterMark( new WriteBufferWaterMark( lowWaterMark, highWaterMark ) );
+        channel.pipeline().addLast( listener );
+    }
+
+    @Override
+    public void acquire( Channel channel )
+    {
+        ThrottleLock lock = channel.attr( LOCK_KEY ).get();
+
+        while ( channel.isOpen() && !channel.isWritable() )
+        {
+            lock.lock( channel, 1000 );
+        }
+    }
+
+    @Override
+    public void release( Channel channel )
+    {
+        if ( channel.isWritable() )
+        {
+            ThrottleLock lock = channel.attr( LOCK_KEY ).get();
+
+            lock.unlock( channel );
+        }
+    }
+
+    @Override
+    public void uninstall( Channel channel )
+    {
+        channel.attr( LOCK_KEY ).set( null );
+    }
+
+    @ChannelHandler.Sharable
+    private class ChannelStatusListener extends ChannelInboundHandlerAdapter
+    {
+
+        @Override
+        public void channelWritabilityChanged( ChannelHandlerContext ctx ) throws Exception
+        {
+            release( ctx.channel() );
```

Comment:
Do we need to check writability here with `Channel#isWritable()` before calling `#release()`?

Reply:
the ```Channel#isWritable``` check is being done in ```#release``` implementation.